crates/binjs_io/src/entropy/dictionary.rs

```diff
@@ -110,99 +110,138 @@ macro_rules! increment_instance_count {
+        self.shared_len
+    }
+
+    /// The number of values representing prelude dictionary entries in this table.
+    pub fn prelude_len(&self) -> usize {
+        self.values.len() - self.shared_len
+    }
+
+    /// The number of values in this table.
+    pub fn len(&self) -> usize {
+        self.values.len()
+    }
+}
+impl<T> LinearTable<T>
 where
     T: Eq + std::hash::Hash + Clone + std::fmt::Debug + Ord,
 {
-    /// Create a new IndexedTable from a list of instances.
+    /// Create a new LinearTable from a list of instances.
     ///
     /// Values appearing in `instances` that have `threshold` instances or less are ignored.
     pub fn new(
         value_to_instances: HashMap<T, FilesContaining>,
         threshold: FilesContaining,
     ) -> Self {
-        debug!(target: "dictionary", "Creating a IndexedTable with a threshold of {}", Into::<usize>::into(threshold));
+        debug!(target: "dictionary", "Creating a LinearTable with a threshold of {}", Into::<usize>::into(threshold));
         let value_to_instances = value_to_instances.into_iter().sorted(); // We sort to enforce traversal order.
         let mut values = Vec::with_capacity(INDEXED_TABLE_INITIAL_CAPACITY);
-        let mut indices = HashMap::with_capacity(INDEXED_TABLE_INITIAL_CAPACITY);
+        let mut refs_by_value = HashMap::with_capacity(INDEXED_TABLE_INITIAL_CAPACITY);
         for (value, instances) in value_to_instances {
-            debug!(target: "dictionary", "Should we add {:?} to the IndexedTable?", value);
+            debug!(target: "dictionary", "Should we add {:?} to the LinearTable ({} instances)?", value, instances);
             if instances <= threshold {
                 // Too few instances, skipping.
                 debug!(target: "dictionary", "Too few instances: {} <= {} for {:?}", instances, threshold, value);
                 continue;
             }
-            let len = Index::Shared(values.len());
+            let len = TableRef::Shared(values.len());
             values.push(value.clone());
-            let prev = indices.insert(value, len);
+            let prev = refs_by_value.insert(value, len);
             assert!(prev.is_none());
         }
-        let result = IndexedTable { values, indices };
-        debug!(target: "dictionary", "Dictionary: IndexedTable contains {:?}", result);
+        let shared_len = values.len();
+        let result = LinearTable {
+            values,
+            refs_by_value,
+            shared_len,
+        };
+        debug!(target: "dictionary", "Dictionary: LinearTable contains {:?}", result);
         result
     }
 
-    /// Create an empty `IndexedTable`.
+    /// Create an empty `LinearTable`.
     pub fn with_capacity(len: usize) -> Self {
-        IndexedTable {
+        LinearTable {
             values: Vec::with_capacity(len),
-            indices: HashMap::with_capacity(len),
+            refs_by_value: HashMap::with_capacity(len),
+            shared_len: 0,
         }
```

Comment:
might be nice to add constructor (new static method?)

Reply:
There are already two of them a few lines below :)